Added Create Environment UI

himanshi-chhabra opened the pull request:
### :nut_and_bolt: Description: What code changed, and why?
We need to add UI for Create Environment

<!-- /!\ Please ensure that you are NOT disclosing any customer information without their consent /!\ -->

### :chains: Related Resources
#4544 

### :+1: Definition of Done
Added create functionality for Environment in which we can create a particular environment.
### :athletic_shoe: How to Build and Test the Change
```
STEP 1
inside the hab studio

[default:/src:0]# build components/automate-ui-devproxy/
[default:/src:0]# start_automate_ui_background
[default:/src:0]# start_all_services

STEP 2
open new window
go to automate UI path

$ cd components/automate-ui
and run the command 

npm run serve:hab
```
To add data https://github.com/chef/automate/blob/master/dev-docs/adding-data/adding_test_data.md#adding-data-to-infra-views
 **PFA video for these changes.**

Steps: 
1. Go To  Infrastructure -> Chef Servers -> click server name-> click on org name -> Environment
2. To see the create button type `FEAT` on-screen and a pop-up will appear you need to toggle the ```Infra server(Remaining)``` to ON and then refresh the page, will see a create button on the right of the environment tab page.

### :white_check_mark: Checklist

- [ ] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [ ] Tests added/updated?
- [ ] Docs added/updated?
- [ ] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### :camera: Screenshots, if applicable
![create_update_1](https://user-images.githubusercontent.com/61003053/110821219-de961000-82b5-11eb-9a00-b172b554d301.png)
![create_update_2](https://user-images.githubusercontent.com/61003053/110821438-143af900-82b6-11eb-8d55-06c96a36421b.png)
![create_en3](https://user-images.githubusercontent.com/61003053/110165756-5f1dc200-7e19-11eb-8574-5a5814eca79f.png)
![create_en4](https://user-images.githubusercontent.com/61003053/110165750-5dec9500-7e19-11eb-9f4e-29e90cd85690.png)

https://user-images.githubusercontent.com/61003053/110165581-2382f800-7e19-11eb-8f75-5122d7bea2f5.mp4

vinay033 reviewed (approved):
Changes look good to me. Please change the screenshot.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.html:
```diff
@@ -0,0 +1,110 @@
+        <app-infra-tab tabTitle="Constraints" [active]="constraintsTab" [disabled]="!firstFormGroup?.valid">
+          <div *ngIf="showConstraint">
+            <app-infra-environment-constraint
+              [constraintKeys]="constraintKeysp"
+              (constraintValues)="constraintItemsHandler($event)"
+              [dynamicArray]="dynamicArrayp"
```
Seems like you were trying to use "p" to differentiate names for both of these parameters here... no need to do that. Plus it makes it confusing to the reader.
```suggestion
              [constraintKeys]="constraintKeys"
              (constraintValues)="constraintItemsHandler($event)"
              [dynamicArray]="dynamicArray"
```
Actually, I would further suggest renaming "dynamicArray" on _both_ sides; that conveys no meaning to the reader. Should convey what is in the array or what it is used for.

himanshi-chhabra replied:
Agree with you, updated the changes.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.html:
```diff
@@ -0,0 +1,110 @@
+          <form [formGroup]="thirdFormGroup">
+            <chef-form-field>
+              <textarea chefInput name="dattr" id="dattr" formControlName="dattr" (change)="onChangeJSON($event)"
+                (keyup)="onChangeJSON($event)" [(ngModel)]="default_attr_value">
```
One of the main design tenets of Automate is one-way data flow, which means do not use the `[(ngModel)]` construct. Please rework.

himanshi-chhabra replied:
yes, I have updated the changes.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.html:
```diff
@@ -0,0 +1,110 @@
+          </form>
+        </app-infra-tab>
+        <app-infra-tab tabTitle="Override Attributes" [active]="overrideTab" [disabled]="!firstFormGroup?.valid">
+          <form [formGroup]="fourthFormGroup">
```
Ah, naming again: and here is where describing the data structure instead of its contents comes back to cause a maintenance issue. 

You have first/second/third/fourth-FormGroups, which describe the position I guess, but why should we care about the position? Rather, what is each form group for? Better names would be DetailsFormGroup, DefaultAttrFormGroup, etc. 

Why did I say it is a maintenance issue? You do not even have a secondFormGroup. 😁

himanshi-chhabra replied:
Agreed, updated the changes.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.spec.ts:
```diff
@@ -0,0 +1,177 @@
+      store = TestBed.inject(Store);
+    });
+
+    it('opening creat model and defaults tab to showing section', () => {
```
Besides the typos could you reword to be more clear. Not sure what this is supposed to mean.
```suggestion
    it('opening create modal and defaults tab to showing section', () => {
```

himanshi-chhabra replied:
Updated the description, it's for when we click on the create environment button the tabs inside the pop-up set to default tab which is `Detail tab`.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.spec.ts:
```diff
@@ -0,0 +1,177 @@
+      expect(component.overrideTab).toBe(false);
+    });
+
+    it('opening create modal resets to empty string', () => {
```
resets what to empty string?

himanshi-chhabra replied:
updated the test description.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.spec.ts:
```diff
@@ -0,0 +1,177 @@
+      expect(component.fourthFormGroup.controls['oattr'].value).toEqual('{}');
+    });
+
+    it('on conflict error, modal remains open and displays conflict error', () => {
```
This is not accurate. It is not validating "and displays conflict error".

himanshi-chhabra replied:
updated the testcase

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.spec.ts:
```diff
@@ -0,0 +1,177 @@
+      expect(component.visible).toBe(true);
+    });
+
+    it('on success, closes modal and adds new environment', () => {
```
You are not confirming that the modal closes--you are actually closing the modal manually in the test here.

himanshi-chhabra replied:
Updated the testcase

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.spec.ts:
```diff
@@ -0,0 +1,177 @@
+
+      store.dispatch(new CreateEnvironmentFailure(error));
+
+      expect(component.creating).toBe(false);
```
An earlier test confirms the visibility of the modal with `component.visible`; here you are using  `component.creating`. Why are they not testing the same property?

himanshi-chhabra replied:
In code, getting the status from the store we are setting the creating to false therefore I have set it to false in spec. Now I have updated the changes.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,336 @@
+          this.store.dispatch(new GetEnvironments(payload));
+          this.creating = false;
+
+          // Close the modal on any error other than conflict and display in banner.
```
More clear to say, e.g.:
```suggestion
          // Close the modal on any other error because it will be displayed in the banner.
```

himanshi-chhabra replied:
Updated the comment.

himanshi-chhabra commented on components/automate-ui/src/app/entities/environments/environment.effects.ts:
```diff
@@ -98,4 +102,30 @@ export class EnvironmentEffects {
+    ofType(EnvironmentActionTypes.CREATE_SUCCESS),
+    map(({ payload: { } }: CreateEnvironmentSuccess) => new CreateNotification({
+      type: Type.info,
+      message: 'Created environment.'
```
In UX design we don't have the resource name in the notification, now I have added the changes.

himanshi-chhabra commented on components/automate-ui/src/app/entities/environments/environment.model.ts:
```diff
@@ -4,13 +4,13 @@ export interface Environment {
+  default_attributes: Object;
+  override_attributes: Object;
+  run_list?: string[];
+  cookbook_versions: any;
```
yes, i have added the correct type for `cookbook_version`.

himanshi-chhabra commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.html:
```diff
@@ -0,0 +1,110 @@
+            <div class="input-margin">
+              <chef-form-field>
+                <label>
+                  <span class="label">Name</span>
```
Added the asterisk for required fields.

himanshi-chhabra commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.html:
```diff
@@ -0,0 +1,110 @@
+        <app-infra-tab tabTitle="Default Attributes" [active]="defaultTab" [disabled]="!firstFormGroup?.valid">
+          <form [formGroup]="thirdFormGroup">
+            <chef-form-field>
+              <textarea chefInput name="dattr" id="dattr" formControlName="dattr" (change)="onChangeJSON($event)"
```
Actually using "d" for default and "o" for override attributes. I have changed it to default and override naming.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.html:
```diff
@@ -0,0 +1,109 @@
+                </label>
+                <chef-error
+                  *ngIf="(detailsFormGroup.get('name').hasError('required') || detailsFormGroup.get('name').hasError('pattern')) && detailsFormGroup.get('name').dirty">
+                  Display Name is required.
```
Must match the text in L13:
```suggestion
                  Name is required.
```

himanshi-chhabra replied:
updated changes

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.html:
```diff
@@ -0,0 +1,109 @@
+            <app-infra-environment-constraint
+              [constraintKeys]="constraintKeys"
+              (constraintValues)="constraintItemsHandler($event)"
+              [dynamicArray]="dynamicArray"
```
As I mentioned last time, I would further suggest renaming "dynamicArray" on both sides; that conveys no meaning to the reader. Should convey what is in the array or what it is used for.

himanshi-chhabra replied:
changed dynamic array to cookbook version name . Updated changes

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.scss:
```diff
@@ -0,0 +1,454 @@
+      }
+
+      ::ng-deep #default::-webkit-scrollbar-thumb {
+        background: #B4C3CC;
```
Avoid color literals -- create a named constant.

himanshi-chhabra replied:
yes i have added this to variable file

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.spec.ts:
```diff
@@ -0,0 +1,179 @@
+
+  describe('create environment', () => {
+    let store: Store<NgrxStateAtom>;
+    const environment = {
```
Can we strongly type this constant? That provides a guarantee that all the properties are valid.

msorens replied:
Not sure why you marked this resolved; it still has no type.

himanshi-chhabra replied:
yes, i have updated the changes

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.spec.ts:
```diff
@@ -0,0 +1,179 @@
+      store = TestBed.inject(Store);
+    });
+
+    it('opening create modal and reset tab to details tab', () => {
```
I think this is a bit more clear.
```suggestion
    it('opening create modal resets default tab to details tab', () => {
```

himanshi-chhabra replied:
updated test line.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.spec.ts:
```diff
@@ -0,0 +1,179 @@
+      store.dispatch(new CreateEnvironmentSuccess({environment}));
+      expect(component.creating).toBe(false);
+      expect(component.visible).toBe(false);
+      expect(component.environmentsList.length).not.toBeNull();
```
This is not confirming what the test purports: that the specific new environment has been added.

himanshi-chhabra replied:
Agreed. :)

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+  public jsonString: string;
+  public defaultAttrParseError: boolean;
+  public overrideAttrParseError: boolean;
+  public data: any;
```
Can we use a strong type here?

himanshi-chhabra replied:
i have updated the onjsonchange function and no need to use this variable.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+      this.org = this.orgId;
+    });
+
+    const payload = {
```
As mentioned in a previous file, constant objects need a strong type so we know all properties are valid.

himanshi-chhabra replied:
added type

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+          this.constraintKeys.push(cookbook.name);
+        });
+      }
+      this.name_id = this.constraintKeys[0];
```
Is it valid to always use the first one? Would be helpful to have an in-code comment explaining that.

himanshi-chhabra replied:
yes, when we click on create environment pop-up then, by default the dropdown value of the constraint name set to be 1st value of constraint keys.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+
+  createEnvironment() {
+    this.creating = true;
+    const environment = {
```
Again, need a specific type on this.

himanshi-chhabra replied:
Added changes

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+      });
+  }
+
+  public handleInput(event: KeyboardEvent): void {
```
Let's be consistent: you use `public` here but not in any of the other methods below.

himanshi-chhabra replied:
yes, i have updated this in all the files.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+  }
+
+  // Getting list of cookbook names
+  loadCookbookConstraint() {
```
```suggestion
  private loadCookbookConstraint() {
```

himanshi-chhabra replied:
updated changes

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+    );
+  }
+
+  toDisplay(cookbookVersions: Array<DynamicGrid> = []) {
```
```suggestion
  private toDisplay(cookbookVersions: Array<DynamicGrid> = []) {
```

himanshi-chhabra replied:
added changes

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+    this.overrideTab = false;
+  }
+
+  ngOnDestroy(): void {
+    this.isDestroyed.next(true);
+    this.isDestroyed.complete();
+  }
```
Suggest move this to be right after `ngOnInit`.

himanshi-chhabra replied:
added

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+    }
+  }
+
+  // registers 'fn' that will be fired wheb changes are made
```
```suggestion
  // registers 'fn' that will be fired when changes are made
```

himanshi-chhabra replied:
added changes

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+    try {
+        // parse it to json
+      this.data = JSON.parse(newValue);
+      if (this.textareaID === 'default') { (this.defaultAttrParseError = false); }
+      if (this.textareaID === 'override') { (this.overrideAttrParseError = false); }
```
Looks like these are unnecessary.

himanshi-chhabra replied:
yes i have updated the code.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+
+  // registers 'fn' that will be fired wheb changes are made
+  // this is how we emit the changes back to the form
+  public registerOnChange(fn: any) {
+    this.propagateChange = fn;
+  }
```
This is unused; should it be called somewhere?

himanshi-chhabra replied:
yes, i have updated the onjsonchange function now it's not required.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,335 @@
+
+    }
+    // update the form
+    this.propagateChange(this.data);
```
Because `registerOnChange` is never called, this call always uses the default, which does nothing.

himanshi-chhabra replied:
updated

msorens commented on components/automate-ui/src/app/modules/infra-proxy/environment-details/environment-details.component.ts:
```diff
@@ -127,7 +129,7 @@ export class EnvironmentDetailsComponent implements OnInit, OnDestroy {
   }
 
-  toDisplay(cookbookVersions: CookbookVersion[]): CookbookVersionDisplay[] {
+  toDisplay(cookbookVersions: Object): CookbookVersionDisplay[] {
```
Why this change?

himanshi-chhabra replied:
while creating the environment or while adding the particular environment we are using cookversion as object type that's why I changed it to object.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/environments/environments.component.ts:
```diff
@@ -20,23 +21,30 @@ export class EnvironmentsComponent implements OnInit, OnDestroy {
     private layoutFacade: LayoutFacadeService
-  ) { }
+  ) {
+    // feature flag enable and disable the create button
+    this.chefInfraViewsFeatureFlagOn =
+    this.featureFlagsService.getFeatureStatus('chefInfraTabsViews');
```
```suggestion
    // feature flag enables or disables the create button
    this.chefInfraViewsFeatureFlagOn =
      this.featureFlagsService.getFeatureStatus('chefInfraTabsViews');
```

himanshi-chhabra replied:
updated the comment

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.html:
```diff
@@ -0,0 +1,77 @@
+    </label>
+    <div class="drop">
+      <mat-select panelClass="chef-dropdown" [(value)]="name_id" [disabled]="!constraintKeys.length">
+        <mat-option *ngFor="let list of constraintKeys" [value]="list" [ngClass]="{'checkmark':list==name_id}">
```
I think you mean this:
```suggestion
        <mat-option *ngFor="let key of constraintKeys" [value]="key" [ngClass]="{'checkmark': key === name_id}">
```
So: 
1. Change `list` to `key`
2. use `===` not `==` 
3. add whitespace

This applies here and in other occurrences in this file.

himanshi-chhabra replied:
updated the changes

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.html:
```diff
@@ -0,0 +1,77 @@
+</div>
+<div class="fixed-height">
+  <div class="scroll">
+    <div class="row-margin" *ngFor="let dynamic of dynamicArray; let i = index;">
```
Really need to give the reader an idea of what a dynamic or dynamicArray is; let's use better names.

himanshi-chhabra replied:
updated the variable name

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,188 @@
+      version: ['', [Validators.required, Validators.pattern(Regex.patterns.VALID_VERSION)]]
+    });
+
+    this.operatorKeys = ['~>', '>=', '>', '=', '<', '<='];
```
Add an in-code comment explaining `~>`

himanshi-chhabra replied:
added code comments.

msorens replied:
Again, you did not read my request. 
(1) Please delete the comment `// operators for Cookbook version`; that is self-evident.
(2) Please add an in-code comment explaining `~>`

himanshi-chhabra replied:
Added in-code comment for ```~>``` :  **```approximately greater than; also known as "pessimistically greater than", or "pessimistic"```**

himanshi-chhabra replied:
Are you looking for this above change?

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,188 @@
+    );
+  }
+
+  selected(value, i) {
```
`i` is not an appropriate name for a function argument; needs to convey some meaning. Perhaps `languageIndex` ?

himanshi-chhabra replied:
added type in all the functions

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,188 @@
+
+  selected(value, i) {
+
+    let previousValue;
```
Need a type for this.

himanshi-chhabra replied:
added the changes

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,188 @@
+    );
+  }
+
+  selected(value, i) {
```
Wow! What is this method doing?
Suggest each block should be extracted into a named function so this function will look something like:
```
selectLanguage(...) {
  setLanguageValueFromIndex(...);
  setNameIn???ArraytoValue(...);
  doSomethingSomethingWithConstraints(...);
  etc...
}
```
And looking further, you can then reuse those sub-functions below in the `handle*` methods.

himanshi-chhabra replied:
working on these changes, will update

himanshi-chhabra replied:
Those **handle** methods are specific to name, operator, version that's why i created separated functions to handle these changes.

msorens replied:
Looks like all you did was rename `selected` to `handleName`, etc., which is not at all what I was asking. Inside `handleEditName` you have what looks like 4 blocks of code. Each of those 4 blocks should be extracted into a named function so `handleEditName` will look something like I showed above. _And then you can reuse those subfunctions in `handleEditOperator` and `handleEditVersion`._

himanshi-chhabra replied:
I have added the functions for each four blocks.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,188 @@
+    this.conflictError = false;
+  }
+
+  // Handle the version change
+  public handleVersion(event, i) {
```
Delete the comment; it does not add any information.
```suggestion
  public handleVersionChange(event, i) {
```

Same thing for `operator` and `name` below.

himanshi-chhabra replied:
added the changes

SEAjamieD reviewed (changes requested):
Similar to my comments on #4813 - in the SCSS files there is a lot of fixed widths and heights being used, which is ultimately causing certain elements to remain stiff in the view.  These views should all be made pretty flexible.

If you'd like more specific guidance on this, I'm happy to go more in depth, or review in a pair session.

msorens commented:
Nice catch, @SEAjamieD !

himanshi-chhabra commented:
> Similar to my comments on #4813 - in the SCSS files there is a lot of fixed widths and heights being used, which is ultimately causing certain elements to remain stiff in the view. These views should all be made pretty flexible.
> 
> If you'd like more specific guidance on this, I'm happy to go more in depth, or review in a pair session.

@SEAjamieD Yes, I would like to have more guidance on this.

himanshi-chhabra commented:
> Similar to my comments on #4813 - in the SCSS files there is a lot of fixed widths and heights being used, which is ultimately causing certain elements to remain stiff in the view. These views should all be made pretty flexible.
> 
> If you'd like more specific guidance on this, I'm happy to go more in depth, or review in a pair session.

@SEAjamieD Hi Jamie, regarding this I have added the fixed-width as we discussed and also added the visual indicator of which button is in focus.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,311 @@
+          this.constraintKeys.push(cookbook.name);
+        });
+      }
+      // first cookbook constrains keys selected on drop-down when loading constrain data
```
```suggestion
      // first cookbook constrains keys selected on drop-down when loading constraint data
```

himanshi-chhabra replied:
Added the changes

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,311 @@
+    }
+  }
+
+  onChangeOverrideJson(event: { target: { value: string } } ) {
+    // get value from text area
+    const newValue = event.target.value;
+    try {
+      // parse it to json
+      JSON.parse(newValue);
+      this.overrideAttrParseError = false;
+    } catch (ex) {
+      // set parse error if it fails
+      this.overrideAttrParseError = true;
```
Comments should only be used for clarification or explication--these are not providing any info to the reader, so are not needed here. I wrote a possibly helpful article on best practices of writing comments: [Fighting Evil in Your Code: Comments on Comments](https://www.red-gate.com/simple-talk/opinion/opinion-pieces/fighting-evil-code-comments-comments/)

himanshi-chhabra replied:
removed these comments

msorens commented on components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.ts:
```diff
@@ -0,0 +1,311 @@
+  }
+
+  tabChange(tab: number) {
+    // Tab indices here correspond with the order of
+    // `<app-infra-tab-change>` elements in the template.
```
By comparison, this comment is important and useful. 😁

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,160 @@
+export class InfraEnvironmentConstraintComponent implements OnInit {
+
+  @Input() constraintKeys: string[] = [];
+  @Input() cookbookConstraintArray: Array<CookbookConstraintGrid> = [];
```
This is both more verbose than it needs to be and also inconsistent. Why should this particular variable include its type in the name ("Array") when all other variables do not? It is generally against best practices because it adds to the maintenance burden (e.g. what if it later became a `List` instead of an `Array`?) 
```suggestion
  @Input() cookbookConstraints: Array<CookbookConstraintGrid> = [];
```

himanshi-chhabra replied:
Updated the variable name

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,160 @@
+    this.constraintValues.emit(this.cookbookConstraintArray);
+  }
+
+  // Deleting a particular CookbookVersion
```
Please delete this self-evident comment.

himanshi-chhabra replied:
Removed the comment

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,160 @@
+  }
+
+  // Deleting a particular CookbookVersion
+  public deleteCookbookVersionw(cookbookIndex: number, cookbookConstraint: CookbookConstraintGrid) {
```
Typo:
```suggestion
  public deleteCookbookVersion(cookbookIndex: number, cookbookConstraint: CookbookConstraintGrid) {
```

himanshi-chhabra replied:
updated the typo

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,160 @@
+    this.constraintKeys.push(cookbookConstraint.name);
+    this.constraintKeys.sort((a, b) => a < b ? -1 : a > b ? 1 : 0);
+
+    this.selectedCookbookNames.forEach((_, index) => {
+      if (cookbookIndex === index) {
+      this.selectedCookbookNames.splice(index, 1);
+
+      }
+    });
```
Simplify to one line:
```suggestion
    this.selectedCookbookNames.splice(cookbookIndex, 1);
```

himanshi-chhabra replied:
added in one line

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,160 @@
+    this.conflictError = false;
+  }
+
+  public handleEditName(newCookbookvalue: string, cookbookIndex: number) {
+    let previousCookbookvalue: string;
```
Be specific: use "name" instead of "value" here. No need to be so verbose with having "cookbook" in there, too; that is clear from the context.
```suggestion
  public handleEditName(newName: string, cookbookIndex: number) {
    let previousName: string;
```

himanshi-chhabra replied:
Updated the changes

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-tab-change/infra-tab-change.component.spec.ts:
```diff
@@ -0,0 +1,25 @@
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
```
Please add a couple real tests, too.

himanshi-chhabra replied:
I have removed this file as it will not specify any to run test case as I already added a test case regarding tabs in Create spec file.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-tab-change/infra-tab/infra-tab.component.ts:
```diff
@@ -0,0 +1,16 @@
+import { Component, Input } from '@angular/core';
+
+@Component({
+  selector: 'app-infra-tab',
+  template: `
+  <div [hidden]="!active" class="test">
+    <ng-content></ng-content>
+  </div>
+`
+})
+// TODO:eng-ex The active tab should be tracked in the parent and not the individual tab
+export class InfraTabComponent {
+  @Input() tabTitle: string;
+  @Input() active = false;
+  @Input() disabled = true;
+}
```
Instead of making an almost exact duplicate of an existing component, please reuse the existing one (`TabComponent`) -- which means add the `disabled` property that you seem to need.

himanshi-chhabra replied:
Yes, I agree with you I have already tried to use the TabComponent file but it's using class="pane" in HTML and if I try to override the CSS of that class it will override at everyplace. that's the reason I have added this file.

msorens replied:
@SEAjamieD Could you take a look at this. Should be able to eliminate the duplication. (OK to do this in a follow-up PR.)

SEAjamieD replied:
Yeah we definitely shouldn't be making new components that are this similar.  Worst case lets extend the component we already have.

I've created a new issue to manage this situation here: https://github.com/chef/automate/issues/4866

msorens commented:
Please also (a) get your unit tests passing in buildkite, and (b) talk to the releng team about getting your PR unblocked in the verify-private pipeline--that seems to be happening every time and you need to get it fixed.

himanshi-chhabra commented:
> Please also (a) get your unit tests passing in buildkite, and (b) talk to the releng team about getting your PR unblocked in the verify-private pipeline--that seems to be happening every time and you need to get it fixed.

@msorens yes, it get resolved

SEAjamieD commented:
![Screen Shot 2021-03-22 at 11 09 16 AM](https://user-images.githubusercontent.com/16737484/112037658-2fe89f80-8aff-11eb-8dd3-4e37074ddd7a.png)

The biggest concern in this PR right now is that none of the buttons above are keyboard accessible.  We should always be making sure that anything our users can do with a mouse, can also be done via keyboard.

SEAjamieD commented:
> > Similar to my comments on #4813 - in the SCSS files there is a lot of fixed widths and heights being used, which is ultimately causing certain elements to remain stiff in the view. These views should all be made pretty flexible.
> > If you'd like more specific guidance on this, I'm happy to go more in depth, or review in a pair session.
> 
> @SEAjamieD Hi Jamie, regarding this I have added the fixed-width of the modal in create Environment as we discussed and also added the visual indicator of which button is in focus.

This will work for now.  We should pair program sometime in the near future and I can be more explicit about what flexible means and how we should strive for flexibility.  I will sync up with UX on the new designs and we can come up with a look for the Modals at small and large screen sizes, and then we can drop the new styles into all of the newly designed modals.

himanshi-chhabra commented:
> ![Screen Shot 2021-03-22 at 11 09 16 AM](https://user-images.githubusercontent.com/16737484/112037658-2fe89f80-8aff-11eb-8dd3-4e37074ddd7a.png)
> 
> The biggest concern in this PR right now is that none of the buttons above are keyboard accessible. We should always be making sure that anything our users can do with a mouse, can also be done via keyboard.

@SEAjamieD i have added the changes

SEAjamieD commented:
I've approved the UI changes, however I think you should wait for @msorens approval before merging

himanshi-chhabra commented:
> > > Similar to my comments on #4813 - in the SCSS files there is a lot of fixed widths and heights being used, which is ultimately causing certain elements to remain stiff in the view. These views should all be made pretty flexible.
> > > If you'd like more specific guidance on this, I'm happy to go more in depth, or review in a pair session.
> > 
> > 
> > @SEAjamieD Hi Jamie, regarding this I have added the fixed-width of the modal in create Environment as we discussed and also added the visual indicator of which button is in focus.
> 
> This will work for now. We should pair program sometime in the near future and I can be more explicit about what flexible means and how we should strive for flexibility. I will sync up with UX on the new designs and we can come up with a look for the Modals at small and large screen sizes, and then we can drop the new styles into all of the newly designed modals.

@SEAjamieD Sure, will create a new issue regarding this.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,148 @@
+  constructor(
+  ) {
+
+    this.operatorKeys = ['~>', // approximately greater than;
+    // also known as "pessimistically greater than", or "pessimistic"
```
I realize you copied this from the docs, but it is still a mystery.
Your job is to explain your code to the reader. What does it mean? This comment is inadequate.

Also, would suggest to rename the variable to `operators` instead of `operatorKeys`, to be less confusing.

msorens commented on components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts:
```diff
@@ -0,0 +1,148 @@
+    this.selectedCookbookNames.forEach((element, index) => {
+      if (index === cookbookIndex) {
+        previousName = element;
+        if (newName && newName !== 'undefined') {
```
Is this really looking for the string "undefined" ?? Or are you trying to check if newName is undefined? If the latter, that will not work. The non-value `undefined` is falsy, so all you need is this:
```suggestion
        if (newName) {
```
Note that if this was unit-tested, I would not have to ask the question. 😁 
And I am just noticing that this file does not have any unit tests; there is a lot going on here and it definitely needs tests.

msorens reviewed (approved):
Looking much better. I would like to urge that this PR be examined for missing unit tests; I point out one file below but there may be others. I would normally use "request changes" because of the lack of tests, but because of deadlines, I am approving on the condition that you will add some unit tests before you merge. 😁

himanshi-chhabra commented:
While acceptance testing, observed

- In Create Environment, disable create button if a name already exists.
- change the border color of the text-field when focused.
will cover in #4873
